packages/docusaurus-theme-common/src/utils/codeBlockUtils.ts

```diff
@@ -211,39 +255,38 @@ export function parseLanguage(className: string): string | undefined {
  */
 export function parseLines(
   content: string,
-  options: {
-    /**
-     * The full metastring, as received from MDX. Line ranges declared here
-     * start at 1.
-     */
-    metastring: string | undefined;
-    /**
-     * Language of the code block, used to determine which kinds of magic
-     * comment styles to enable.
-     */
-    language: string | undefined;
-    /**
-     * Magic comment types that we should try to parse. Each entry would
-     * correspond to one class name to apply to each line.
-     */
-    magicComments: MagicCommentConfig[];
-  },
-): {
-  /**
-   * The highlighted lines, 0-indexed. e.g. `{ 0: ["highlight", "sample"] }`
-   * means the 1st line should have `highlight` and `sample` as class names.
-   */
-  lineClassNames: {[lineIndex: number]: string[]};
-  /**
-   * If there's number range declared in the metastring, the code block is
-   * returned as-is (no parsing); otherwise, this is the clean code with all
-   * magic comments stripped away.
-   */
-  code: string;
-} {
-  let code = content.replace(/\n$/, '');
-  const {language, magicComments, metastring} = options;
+  options: CodeBlockParseLinesOptions,
+): CodeBlockParsedLines {
+  const parsedLines: CodeBlockParsedLines = {
+    code: content.replace(/\n$/, ''),
+    lineClassNames: {},
+  };
+
   // Highlighted lines specified in props: don't parse the content
+  if (fillLineClassNamesFromMetastring(parsedLines, options)) {
+    return parsedLines;
+  }
+
+  fillLineClassNamesFromCode(parsedLines, options);
+
+  return parsedLines;
```

Comment:
Pseudo code but I expect it to look more like this:

```suggestion
  return getLineClassNamesFromMetastring(options) ?? getLineClassNamesFromCode(code, options);
```